### PR TITLE
Parser seperation

### PR DESCRIPTION
**Pull request by MengnanLi91**
This PR is designed to separate current MOOSE parser into a front part and end part. The front part will parse before MooseApp is built. 

## Reason

This work is part of the NEAMS task: Application block to host: dynamic loading params

## Design

The plan is to separate the part of parser which is not related to MooseApp as front parser and keep the rest as the new parser.

## Impact

This will help the new capability Application block to host

### CODE REVIEW

**Comment by MengnanLi91** [comment]
@loganharbour I've done all the changes I could at this point. Marmot and Griffin failure will need to be patched on app side given the error message(I will talk to Patrick about all the app patches after this PR has been merged) and I need some help with MooseServerTest failure. Could you review this PR when you have time?

File: examples/ex01_inputfile/src/main.C
Comment (loganharbour) [inline examples/ex01_inputfile/src/main.C:23]: @roystgnr can we move this around?

File: framework/include/parser/Builder.h
Comment (loganharbour) [inline framework/include/parser/Builder.h:54]:
```suggestion
/**
 * Parses MOOSE input using HIT/WASP.
 */
```

File: framework/include/base/MooseApp.h
Comment (loganharbour) [inline framework/include/base/MooseApp.h:358]: Can we move this to the source file and add a `mooseDeprecated()` message? Something like

> MooseApp::parser() is deprecated, use MooseApp::builder() instead.

File: framework/src/base/AppFactory.C
Comment (loganharbour) [inline framework/src/base/AppFactory.C:75]:
```suggestion
  mooseDeprecated("Please update your main.C to adapt new main function in MOOSE framework, see "
                  "test/src/main.C in MOOSE as an example of moose::main() ");
```

File: framework/src/base/MooseApp.C
Comment (loganharbour) [inline framework/src/base/MooseApp.C:966]: Did this error not exist before?

File: framework/src/base/MooseMain.C
Comment (loganharbour) [inline framework/src/base/MooseMain.C:22]: I think the common namespace is `Moose`, not `moose`

File: framework/src/base/MooseServer.C
Comment (loganharbour) [inline framework/src/base/MooseServer.C:275]: Why `=` and `+=` ?

File: test/tests/misc/app_name/tests
Comment (loganharbour) [inline test/tests/misc/app_name/tests:6]: Why all of the changes from

```
input = 'IGNORED'
```

to

```
input = ''
```

?

File: test/tests/parser/param_duplicate_check/tests
Comment (loganharbour) [inline test/tests/parser/param_duplicate_check/tests:4]: What new capability was added to require this test? OR is this just adding coverage that didn't exist before?

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: framework/src/base/MooseApp.C
Comment (MengnanLi91) [reply framework/src/base/MooseApp.C:966]: This error used to be in Parser.C
```
if (_input_filenames.empty())
    mooseError("No input files specified.");
```
I moved it here because we want to allow empty input file or wrong input file type(e.g. input.e) in front parser and deal with all these options later in MooseApp.C, otherwise we will need add a lot of parameters like 
`"--show-outputs"`, `"--mesh-only [mesh_file_name]"`,` "--definition" `, etc. to new Parser along with the condition logic

File: framework/src/base/MooseServer.C
Comment (MengnanLi91) [reply framework/src/base/MooseServer.C:275]: Good catch!

File: test/tests/misc/app_name/tests
Comment (MengnanLi91) [reply test/tests/misc/app_name/tests:6]: This is suggested by @permcody.  In MooseApp.C we have a very long `if elseif and else` condition to deal with different command line options. When we want to ignore input files, however we actually don't want to use 'IGNORED' as a key word to detect "ignore input file condition", it just fells into the else condition in the very end. So we decide to replace the 'IGNORED' key word in test with ''. Therefore the logic below can work, otherwise it will looking for input file named 'IGNORED' in the folder.
```
  input_param.addCommandLineParam<std::vector<std::string>>(
      "input_file",
      "-i <input_files>",
      "Specify one or multiple input files. Multiple files get merged into a single simulation "
      "input.");

  command_line->addCommandLineOptionsFromParams(input_param);

  std::vector<std::string> input_filename;
  command_line->search("input_file", input_filename);

  if (!input_filename.empty())
    front_parser->parse(input_filename);
```

File: test/tests/parser/param_duplicate_check/tests
Comment (MengnanLi91) [reply test/tests/parser/param_duplicate_check/tests:4]: This is more like a patch to the existing duplication check. Joshua found the duplication check in parser doesn't work properly when multiple input files are merged together(only the last input file get checked). I changed the sequence of duplication and file merge to fix it. Probably should start a new PR on this, but it's very simple so I throw it here.

File: examples/ex01_inputfile/src/main.C
Comment (MengnanLi91) [reply examples/ex01_inputfile/src/main.C:23]: I moved `perf_log()` into `main()`, it compiles fine but failed all the tests as below. Does this a registration error?
`symbol not found in flat namespace '__ZN5Moose8perf_logE'`
Here is where I moved `perf_log()`, did I do anything wrong?
```
template <typename DefaultAppType>
void
main(int argc, char * argv[])
{
  MooseInit init(argc, argv);

  DefaultAppType::registerApps();

  const auto default_app_name = MooseUtils::prettyCppType<DefaultAppType>();

  // Create a performance log
  PerfLog perf_log(default_app_name);

  auto app = createMooseApp(default_app_name, argc, argv);

  app->run();
}
```

File: test/tests/misc/app_name/tests
Comment (loganharbour) [reply test/tests/misc/app_name/tests:6]: Works for me.

**Comment by loganharbour** [comment]
Looks like there's a few things still sticking around. did you fix and just not resolve?

**Comment by MengnanLi91** [comment]
> Looks like there's a few things still sticking around. did you fix and just not resolve?

Yes, I have fixed those but not sure if I should resolve before you take a look

**Comment by brandonlangley** [comment]
@MengnanLi91 @loganharbour -

I saw this and was curious about the `MooseServerTest` seg-fault failure so I took a look into what might be happening.

I found that there are two issues with how the `MooseServer` is interacting with this `Parser` / `Builder` split right now.

There are likely other ways to fix these, but I wanted to share my findings and minimal change fixes in case it's helpful.

- - - - -

**Issue <span>#</span>1:**

  - There is an expectation here that the `Parser` already parsed the input at the time the `Builder` gets constructed.

    This is because `Builder` sets its private `_root` member using `_root(parser.getRootNode())` in the constructor.

    Then the `_root` never changes so it is invalid if the `Parser` will parse the input after the `Builder` is constructed.

    This invalid `_root` is later used throughout `Builder` processing and accessed by the `MooseServer` downstream.

  - All these other cases work since their `Parser` does parse the input prior to `MooseApp` and `Builder` construction:

     - `framework/src/multiapps/MultiApp.C` - `createApp()`

     - `framework/src/base/MooseMain.C` - `createMooseApp()`

     - `framework/src/base/AppFactory.C` - `createAppShared()`

    But `MooseServer` does not do `front_parser->parse()` before the `AppFactory::instance().createShared()` call.

**Fix for <span>#</span>1:**

  - Add a `setRootNode()` method to `Builder` that allows its `_root` member to be updated after its construction like:

    ```cpp
    void setRootNode(hit::Node * root) { _root = root; }
    ```

    And call it in the `MooseServer` case in `MooseApp` to set the `_builder` root after the parse but before the build like:

    ```cpp
    _parser->parse(_input_filenames, getParam<std::string>("_input_text"));
    _builder.setRootNode(_parser->getRootNode());
    _builder.builder();
    ```

    Then the `_root` member of `_builder` will be valid at the necessary time of processing and `MooseServer` access.

- - - - -

**Issue <span>#</span>2:**

  - The `MooseServer` got updated here so that it accesses the `_root` member of `Builder` rather than `Parser` with:

    ```cpp
    _check_app->builder().root()->getNodeView()
    ```

    Separating the input processing into sequential stages leads to a problematic scenario with this access to the root.

    If the first `Parser` stage fails due to an input check error, then the `_root` member of the `Builder` never gets set.

    So `MooseServer` cannot use the `_root` member of `Builder` if any input error is reported from the `Parser` stage.

    This currently prevents features such as autocompletion and hierarchal tree creation from working for these cases.

**Fix for <span>#</span>2:**

  - Update the `Builder::root()` method so that it returns `_parser.getRootNode()` instead its own `_root` member.

    Note `Parser::getRootNode()` currently calls `_root.release()` which releases this root from `Parser` ownership.

    Change the `getRootNode()` return from `_root.release()` to `_root.get()` so the `Parser` keeps the ownership.

    This works because `Builder` already owns a `Parser` reference keeping its `_root` in scope for the `Builder` life.

    The `Builder` destructor should also get updated to be empty since will not be handling the `_root` deletion now.

- - - - -

Here is my whole `git diff` change set, and other paths may also fix the issues but hopefully the diagnosis is useful.

```diff
diff --git a/framework/include/parser/Builder.h b/framework/include/parser/Builder.h
index 996f3c0906..4e57f688ee 100644
--- a/framework/include/parser/Builder.h
+++ b/framework/include/parser/Builder.h
@@ -114,7 +114,9 @@ public:
    */
   std::string hitCLIFilter(std::string appname, const std::vector<std::string> & argv);
 
-  hit::Node * root() { return _root; };
+  hit::Node * root() { return _parser.getRootNode(); };
+
+  void setRootNode(hit::Node * root) { _root = root; }
 
 protected:
   /**
diff --git a/framework/include/parser/Parser.h b/framework/include/parser/Parser.h
index aaa05c170f..5b32ead643 100644
--- a/framework/include/parser/Parser.h
+++ b/framework/include/parser/Parser.h
@@ -123,7 +123,7 @@ public:
   {
     // normal input file is provided, front parser is set to parse input file
     if (_root)
-      return _root.release();
+      return _root.get();
     // no valid input file is provided, front parser is skipped
     else
       return nullptr;
diff --git a/framework/src/base/MooseApp.C b/framework/src/base/MooseApp.C
index 407f315403..ffb983289b 100644
--- a/framework/src/base/MooseApp.C
+++ b/framework/src/base/MooseApp.C
@@ -969,6 +969,7 @@ MooseApp::setupOptions()
     else
     {
       _parser->parse(_input_filenames, getParam<std::string>("_input_text"));
+      _builder.setRootNode(_parser->getRootNode());
       _builder.builder();
     }
     if (isParamValid("mesh_only"))
diff --git a/framework/src/parser/Builder.C b/framework/src/parser/Builder.C
index 4f128f2578..528bb70556 100644
--- a/framework/src/parser/Builder.C
+++ b/framework/src/parser/Builder.C
@@ -134,7 +134,7 @@ Builder::Builder(MooseApp & app, ActionWarehouse & action_wh, Parser & parser)
   _input_filenames = _parser.getInputFileNames();
 }
 
-Builder::~Builder() { delete _root; }
+Builder::~Builder() {}
 
 InputParameters
 Builder::validParams()
```

**Comment by loganharbour** [comment]
@brandonlangley the idea of having `Builder::root()` only sometimes return the root that is set seems dangerous and like a poor design

**Comment by MengnanLi91** [comment]
@brandonlangley Thanks a lot for the suggestions! I've adapted the changes but it still failed MooseServerTest.DocumentChangeAndDiagnostics

**Comment by brandonlangley** [comment]
@MengnanLi91

>  I've adapted the changes but it still failed MooseServerTest.DocumentChangeAndDiagnostics

That's because you removed the trailing space from these two diagnostic messages in `MooseServerTest.C` on this PR.

Your editor likely did this automatically. Those two space characters are annoying and I'm going to fix that in another PR.

```diff
TEST_F(MooseServerTest, DocumentChangeAndDiagnostics)
{
  . . .
  std::string diagnostics_expect = R"INPUT(
line:18 column:0 - (BCs/all/boundary):
line:18 column:0 -     the following side sets (ids) do not exist on the mesh: top (2), bottom (3)
- line:18 column:0 - MOOSE distinguishes between "node sets" and "side sets" depending on whether 
- line:18 column:0 - you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding 
+ line:18 column:0 - MOOSE distinguishes between "node sets" and "side sets" depending on whether
+ line:18 column:0 - you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding
line:18 column:0 -     to your side sets are constructed for you by default.
line:18 column:0 -     Try setting "Mesh/construct_side_list_from_node_list=true" if you see this error.
line:18 column:0 -     Note: If you are running with adaptivity you should prefer using side sets.
)INPUT";
  EXPECT_EQ(diagnostics_expect, "\n" + diagnostics_actual.str());
}
```

**Comment by brandonlangley** [comment]
@loganharbour

> the idea of having Builder::root() only sometimes return the root that is set seems dangerous and like a poor design

I'm not sure I understand. That is exactly what is happening with the current design before the update from my [comment](https://github.com/idaholab/moose/pull/25994#issuecomment-1853234892).

```cpp
hit::Node * Builder::root() { return _root; };
```

Thats why, as described in the **Issue <span>#</span>2** part of my comment, this call had been failing with a seg-fault as seen in testing.

```cpp
_check_app->builder().root()->getNodeView()
```

Because in any case where the `Parser` first stage fails due an input check,  the `_root `member of `Builder` is never set.

So currently, `Builder::root()` only sometimes returns the root.

Unless I'm misunderstanding something, the update that I used actually fixes the call so it will **_always_** return the root with:

```cpp
hit::Node * root() { return _parser.getRootNode(); };
```

And as I mentioned in my comment, there definitely could be some other approach to fixing this which is more preferred.

I mostly wanted to help in understanding what is happening and thought I'd also share how I fixed it in case it was useful.

**Comment by loganharbour** [comment]
Sorry @brandonlangley, I should have clarified. This is what confuses me:

```
hit::Node * root() { return _parser.getRootNode(); };
void setRootNode(hit::Node * root) { _root = root; }
```

I don't understand why both of these methods would be necessary. If `root()` is always correct, I don't follow why we would want to also have `setRootNode()`. Having `_root != root()` is what is concerning to me

**Comment by brandonlangley** [comment]
> Sorry @brandonlangley, I should have clarified. This is what confuses me:
> 
> ```
> hit::Node * root() { return _parser.getRootNode(); };
> void setRootNode(hit::Node * root) { _root = root; }
> ```
> 
> I don't understand why both of these methods would be necessary. If `root()` is always correct, I don't follow why we would want to also have `setRootNode()`. Having `_root != root()` is what is concerning to me

@loganharbour -

Yes we're on the same page, that makes complete sense and I also don't think both of those methods should be necessary.

I actually don't think the `Builder` class needs its own private `_root` member at all, therefore `setRootNode()` isn't needed.

The "true" `_root` is the `std::unique_ptr<hit::Node>` in the `Parser` which is why I suggested for the `Builder` call to be:

```
hit::Node * Builder::root() { return _parser.getRootNode(); };
```

while also letting `Parser` keep `_root` ownership and return `_root.get()` instead of `_root.release()` in `getRootNode()`.

The update that I shared was just the most minimal change set that fixed the issues in order to demonstrate what they were.

I too thought about removing the `_root` member from the `Builder` which removes the need for `Builder::setRootNode()`.

And then just let all usages of `_root` in the `Builder` processing logic go through that `Builder::root()` interface instead.

But this would have required updating like 70 references in `Builder` to `_root` and the change set might be less digestible.

So like I said, I figured my five-line fix would help demonstrate the issue and then you guys could fix it however you wanted.

But yeah, we're in agreement that `setRootNode()` should not be needed if `Builder` uses `root()` and not `_root` directly.

**Comment by brandonlangley** [comment]
@MengnanLi91 @loganharbour -

Three other minor suggestions came to mind as I looked through this, so I thought I would share but none are necessary.

1. Both `Parser.h` and `Builder.h` can remove the `friend class MooseServer` lines since those are no longer needed.

2. Rename `Parser::getRootNode()` to `Parser::root()` for consistency with `Builder::root()` as the use is the same.

3. Rename `Builder::builder()` (that runs the build) to `Builder::build()` as it is an action like the `parse()` method.

    - prevents mixing up with the `Builder::Builder()` constructor or `MooseApp::builder()` that returns a reference

File: examples/ex01_inputfile/src/main.C
Comment (roystgnr) [reply examples/ex01_inputfile/src/main.C:23]: Sorry, missed the notification somehow.

That linker error looks like the definition of `perf_log` has been moved but there's still a *declaration* of `perf_log` (possibly one being used!) which hasn't.

**Comment by loganharbour** [comment]
@brandonlangley FYI, cleaned up `MooseServer` some in https://github.com/idaholab/moose/pull/25994/commits/7e22b7ebd83b92708be7a44569b371f70001891a

**Review by loganharbour — CHANGES_REQUESTED** [review]
Few small changes, did a full review. Should be good to go after these and the app fixes.

**Comment by MengnanLi91** [comment]
> Few small changes, did a full review. Should be good to go after these and the app fixes.

I've address the comments. App fixes should be the only thing that left

**Comment by brandonlangley** [comment]
@loganharbour - Should this reference any actual issue number separate from a commit tagging its own PR number?

**Comment by loganharbour** [comment]
## Preemptive patches

- [ ] Griffin https://github.inl.gov/ncrc/griffin/pull/1691

## Patches for after

- [ ] Griffin https://github.inl.gov/ncrc/griffin/pull/1692

**Comment by loganharbour** [comment]
All patches are in. waiting for devel -> master.

**Review by loganharbour — APPROVED** [review]

**Review by milljm — APPROVED** [review]

**Review by permcody — APPROVED** [review]
Looks good Mengnan - nice work!

**Pull request merged**
